bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/handler/SmartHomeDeviceHandler.java

```diff
@@ -0,0 +1,391 @@
+import org.openhab.binding.amazonechocontrol.internal.smarthome.HandlerBase.UpdateChannelResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+/**
+ * @author Lukas Knoeller - Initial contribution
+ */
+@NonNullByDefault
+public class SmartHomeDeviceHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(SmartHomeDeviceHandler.class);
+
+    private @Nullable SmartHomeBaseDevice smartHomeBaseDevice;
+    private final Gson gson;
+    private final Map<String, @Nullable HandlerBase> handlers = new HashMap<>();
+    private final Map<String, @Nullable JsonArray> lastStates = new HashMap<>();
+
+    public SmartHomeDeviceHandler(Thing thing, Gson gson) {
+        super(thing);
+        this.gson = gson;
+    }
+
+    public synchronized void setDeviceAndUpdateThingState(AccountHandler accountHandler,
+            @Nullable SmartHomeBaseDevice smartHomeBaseDevice) {
+        if (smartHomeBaseDevice == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Can't find smarthomeBaseDevice");
+            return;
+        }
+        boolean changed = this.smartHomeBaseDevice == null;
+        this.smartHomeBaseDevice = smartHomeBaseDevice;
+
+        Set<String> unusedChannels = new HashSet<>();
+        thing.getChannels().forEach(channel -> unusedChannels.add(channel.getUID().getId()));
+
+        Set<String> unusedHandlers = new HashSet<>(handlers.keySet());
+
+        Map<String, List<SmartHomeCapability>> capabilities = new HashMap<>();
+        getCapabilities(capabilities, accountHandler, smartHomeBaseDevice);
+
+        for (String interfaceName : capabilities.keySet()) {
+            HandlerBase handler = handlers.get(interfaceName);
+            if (handler != null) {
+                unusedHandlers.remove(interfaceName);
+            } else {
+                Function<String, HandlerBase> creator = Constants.HANDLER_FACTORY.get(interfaceName);
+                if (creator != null) {
+                    handler = creator.apply(interfaceName);
+                    handlers.put(interfaceName, handler);
+                }
+            }
+            if (handler != null) {
+                Collection<ChannelInfo> required = handler.initialize(this, capabilities.get(interfaceName));
+                for (ChannelInfo channelInfo : required) {
+                    unusedChannels.remove(channelInfo.channelId);
+                    if (addChannelToDevice(channelInfo.channelId, channelInfo.itemType, channelInfo.channelTypeUID)) {
+                        changed = true;
+                    }
+                }
+            }
+        }
+
+        unusedHandlers.forEach(handlers::remove);
+        if (!unusedChannels.isEmpty()) {
+            changed = true;
+            unusedChannels.forEach(this::removeChannelFromDevice);
```

Comment:
This is highly inefficient, each iteration of the loop results in a fresh edit of the Thing and thus triggers a lot of events that could be avoided if you just called `editThing` once, removed all the unwanted channels from the builder, then called `updateThing`.

Reply:
done.